llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4831,18 +4831,38 @@ Instruction *InstCombinerImpl::visitFreeze(FreezeInst &I) {
     return BestValue;
   };
 
+  auto getValueReplacement = [&I, &AC = this->AC, &DT = this->DT]() -> Value * {
+    Value *BestValue = nullptr;
+    for (const auto *U : I.users()) {
+      Value *C = nullptr;
+      Value *SelectArgument = nullptr;
+      if (match(U, m_c_Select(m_Specific(&I), m_Value(SelectArgument))) &&
+          isGuaranteedNotToBeUndefOrPoison(SelectArgument, &AC, &I, &DT))
+        C = SelectArgument;
+
+      if (!BestValue)
+        BestValue = C;
+      else if (BestValue != C)
+        return nullptr;
+    }
+    return BestValue;
+  };
```

Comment:
Splitting the value and constant cases is also not ideal, because values can be constants. So e.g. if we have a `select i1 %c, i32 -1, i32 %freeze` and a `or i32 %freeze, %y` then the best choice would be to use `-1`, which allows both the select and the or to fold. We'd do that if both cases are combined, but not if they are separated.

Reply:
Moved value and constant cases back together.